modules/battery-statefs.c

```diff
@@ -0,0 +1,1329 @@
+static void     mcebat_update_schedule    (void);
+
+/* ------------------------------------------------------------------------- *
+ * TRACKER  --  binds statefs file to sfsbat_t member
+ * ------------------------------------------------------------------------- */
+
+typedef struct tracker_t tracker_t;
+
+/** Bind statefs file to member of sfsbat_t structure */
+struct tracker_t
+{
+    /** Basename of the input file */
+    const char *name;
+
+    /** Path to input file, set at tracker_init() / sfsctl_init() */
+    char       *path;
+
+    /** Pointer to a sfsbat_t member */
+    void       *value;
+
+    /** Value type specific input parser hook */
+    bool      (*update_cb)(tracker_t *, const char *);
+
+    /** File descriptor for the input file */
+    int         fd;
+
+    /** For use with debugging with pipes instead of real statefs */
+    bool        seekable;
+};
+
+static const char *tracker_propdir(void);
+
+static void tracker_init        (tracker_t *self);
+static void tracker_quit        (tracker_t *self);
+static bool tracker_open        (tracker_t *self);
+static void tracker_close       (tracker_t *self);
+static bool tracker_start       (tracker_t *self);
+static bool tracker_read_data   (tracker_t *self, char *data, size_t size);
+static bool tracker_parse_int   (tracker_t *self, const char *data);
+static bool tracker_parse_bool  (tracker_t *self, const char *data);
+static bool tracker_update      (tracker_t *self);
+
+/* ------------------------------------------------------------------------- *
+ * SFSCTL  --  controls for statefs tracking
+ * ------------------------------------------------------------------------- */
+
+static void     sfsctl_init             (void);
+static void     sfsctl_quit             (void);
+
+static void     sfsctl_start            (void);
+static bool     sfsctl_start_try                (void);
+static gboolean sfsctl_start_cb         (gpointer aptr);
+
+static bool     sfsctl_watch_cb         (struct epoll_event *eve, int cnt);
+
+static void     sfsctl_schedule_reread  (void);
+static void     sfsctl_cancel_reread    (void);
+static gboolean sfsctl_reread_cb        (gpointer aptr);
+
+/* ------------------------------------------------------------------------- *
+ * MODULE_INIT_EXIT
+ * ------------------------------------------------------------------------- */
+
+G_MODULE_EXPORT const gchar *g_module_check_init (GModule *module);
+G_MODULE_EXPORT void         g_module_unload     (GModule *module);
+
+/* ========================================================================= *
+ * MISC_UTILS
+ * ========================================================================= */
+
+/** Lookup table for sfsbat_state_t parsing */
+static const struct {
+    /** State name visible in statefs file */
+    const char     *name;
+
+    /** Enumeration ID used in code */
+    sfsbat_state_t  state;
+} state_lut[] =
+{
+    { "charging",    STATEFS_BATTERY_STATE_CHARGING    },
```

Comment:
On the other hand you are right: maybe "unknown" should be anything except known values. Also empty string is suitable for this state.